### DISCUSSION TITLE
How to extrapolate a spectrum linearly?

### DISCUSSION THREAD

**Discussion by frankkjensen** [colour-science/colour#1106] [Q&A] [answered]
Hi,

I am trying to extrapolate some spectra. As default they will be extrapolated using 'constant', i.e. will just continue with the last value. I try to set the extrapolator method to 'Linear', but it doesn't work. What am I doing wrong?

```
sd.extrapolator_kwargs['method'] = 'Linear'
sd_reshaped = colour.colorimetry.reshape_sd(sd, colour.SpectralShape(300, 900, 1))
```

Best regards,
Frank

**Comment by thomasmansencal** [comment, marked as answer]
Hi @frankkjensen,

This one is confusing and I think we should document it better! There are two schools for interpolating and extrapolating the spectral data in Colour:

- The _Old School - CIE Compliant_ way that uses the `interpolate`, `extrapolate`, `align` (and `trim`) methods of `colour.SpectralDistribution`, those will follow CIE recommendations.
- The _New School - Ignore CIE_ that leverages the underlying continuous function that allows you to do things like that: `sd[colour.SpectralShape(300, 900, 1).wavelengths]` and directly sample it.

What you did by changing directly the `extrapolator_kwargs` would have worked with the latter. Unfortunately for the `colour.colorimetry.reshape_sd` definition that uses the _Old School - CIE Compliant_ way, you need to pass kwargs as follows:
`colour.colorimetry.reshape_sd(sd, colour.SpectralShape(300, 900, 1), extrapolator_kwargs={"method": "Linear"})`. The reason is that the *CIE 15:2004* and *CIE 167:2005* recommend constant extrapolation [so it is hardcoded](https://github.com/colour-science/colour/blob/fb6f7aa466e2a9b6c3ea4e4ad98fedbf8b789eb0/colour/colorimetry/spectrum.py#L1302) but can still be overridden.

Cheers,

Thomas

**Comment by frankkjensen** [comment]
Ok, thanks for the explanation :-)

Best regards,
Frank
